finality-aleph/src/sync/handler/mod.rs

```diff
@@ -708,6 +718,137 @@ mod tests {
+            assert!(maybe_id.is_none(), "should not import justification");
+            assert!(maybe_error.is_none(), "should work");
+            let mut counter = 0;
+            while let Ok(BlockImported(header)) = notifier.next().await {
+                handler.block_imported(header).expect("should work");
+                counter += 1;
+                if counter == branch.len() {
+                    break;
+                }
+            }
+            // increasingly larger gaps
+            let partial = branch.clone()[session as usize + 1..].to_vec();
+            if partial.is_empty() {
+                break;
+            }
+            let response_items = branch_response(partial.clone(), false, false, true);
+            let (maybe_id, maybe_error) = handler.handle_request_response(response_items, peer_id);
+            assert!(maybe_id.is_some(), "should import justification");
+            assert!(maybe_error.is_none(), "should work");
+            // get notification about finalized end-of-session block
+            match notifier.next().await.expect("should not be closed") {
+                BlockFinalized(header) => {
+                    assert_eq!(header.id().number(), top, "should finalize the top block")
+                }
+                _ => panic!("should notify about finalized block"),
+            };
+        }
+    }
+
+    #[tokio::test]
+    async fn skips_justification_gap_with_last_of_current_session_only() {
+        let (mut handler, _backend, mut notifier, genesis) = setup();
+        let last_block_of_first_session = SESSION_BOUNDARY_INFO.last_block_of_session(SessionId(0));
+        let last_block_of_second_session =
+            SESSION_BOUNDARY_INFO.last_block_of_session(SessionId(1));
+        let peer_id = 0;
+        let branch_low = grow_light_branch_till(
+            &mut handler,
+            &genesis,
+            &last_block_of_first_session,
+            peer_id,
+        );
+        let finalizing_justification =
+            MockJustification::for_header(branch_low.last().expect("should not be empty").clone());
+        let branch_high = grow_light_branch_till(
+            &mut handler,
+            &finalizing_justification.id(),
+            &last_block_of_second_session,
+            peer_id,
+        );
+        let response_items = branch_response(
+            branch_low
+                .iter()
+                .cloned()
+                .chain(branch_high.iter().cloned())
+                .collect(),
+            true,
+            true,
+            false,
+        );
+        let (maybe_id, maybe_error) = handler.handle_request_response(response_items, peer_id);
+        assert!(maybe_id.is_none(), "should not import justification");
+        assert!(maybe_error.is_none(), "should work");
+        let mut counter = 0;
+        while let Ok(BlockImported(header)) = notifier.next().await {
```

Comment:
These `while`s could probably be abstracted into a function like `ensure_branch_imported` (called twice here ofc).

Reply:
done